### ISSUE TITLE
vide.d:http fails to build under gdc

### ISSUE THREAD

**Issue by baruch** [rejectedsoftware/vibe.d#1602]
I'm trying to build on Debian with msgpack-rpc that depends on vibe.d and during compile I get the error:

```
vibe-d:http 0.7.29: building configuration "library"...
/usr/lib/gvibe-d:http 0.7.29: building configuration "library"...
/usr/lib/gcc/x86_64-linux-gnu/6/include/d/core/atomic.d:1464:13: error: static assert  "Invalid template type specified."
             static assert(0, "Invalid template type specified.");
             ^
../../../../.dub/packages/vibe-d-0.7.29/vibe-d/source/vibe/http/server.d:1388:51: note: instantiated from here: atomicLoad!(cast(MemoryOrder)5, shared(HTTPServerContext)[])
    else return cast(HTTPServerContext[])atomicLoad(g_contexts);
cc/x86_64-linux-gnu/6/include/d/core/atomic.d:1464:13: error: static assert  "Invalid template type specified."
             static assert(0, "Invalid template type specified.");
             ^
../../../../.dub/packages/vibe-d-0.7.29/vibe-d/source/vibe/http/server.d:1388:51: note: instantiated from here: atomicLoad!(cast(MemoryOrder)5, shared(HTTPServerContext)[])
    else return cast(HTTPServerContext[])atomicLoad(g_contexts);
```

The compiler I happen to use is gdc 6.1.1 (Debian version is 4:6.1.1-1)

**Comment by Scellow** [comment]
Same for me with dmd #1611

**Comment by jpf91** [comment]
ping @s-ludwig 

#1611 is unrelated. The GDC failure is a combination of GDC not supporting 128bit atomic loads and vibe.d not properly checking for support.

The `atomicLoad`, `atomicStore` and `cas` functions only support types up to `size_t.sizeof`. Supporting `2*size_t.sizeof` is a compiler and architecture specific extension and vibe.d should check `has128BitCAS` and `has64BitCAS` before using these extended methods. Note that this in theory also applies when using `atomicLoad(long)` on 32 bit targets. GDC currently supports 64bit on all targets, so this error is only visible on 64bit machines.

See also: https://github.com/D-Programming-GDC/GDC/pull/263 but it's probably not a simple task to support 128bit atomics in GDC as the GCC support for that requires the external `libatomic` library.

**Comment by s-ludwig** [comment]
I closed this without verifying that GDC now actually compiles. If there are more issues, please reopen.

**Comment by jpf91** [comment]
> I closed this without verifying that GDC now actually compiles. If there are more issues, please reopen.

We've also added support for 128bit atomics to GDC. So 0.7.30 should compile fine.

However, IIRC vibe.D master now uses std.experimental.allocator which is not available in the GDC frontend `DMD 2.068.2 `. The overview page still claims support for `2.068.2`: http://code.dlang.org/packages/vibe-d is that information out of date?

**Comment by s-ludwig** [comment]
> We've also added support for 128bit atomics to GDC. So 0.7.30 should compile fine.

Nice, that will definitely be useful!

> However, IIRC vibe.D master now uses std.experimental.allocator which is not available in the GDC frontend DMD 2.068.2. The overview page still claims support for 2.068.2: http://code.dlang.org/packages/vibe-d is that information out of date?

I've updated README.md on the master branch to match the supported compilers for 0.8.0 and up. 0.7.31 will still support 2.068.2, but 0.8.0, which is supposed to be released in parallel will only work starting with 2.069.x. It's possible that there will be another 0.7.x release afterwards, should 0.8.0 turn out to be problematic w.r.t. upgrading existing applications., so there is still some room/time for active maintenance for 2.068.2
